### PR TITLE
refactor(core): add support to multiple webviews on a Tauri window

### PR DESCRIPTION
**Pull request by lucasfernog**
### What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Docs
- [ ] New Binding issue #___
- [ ] Code style update
- [x] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

### Does this PR introduce a breaking change?

- [ ] Yes, and the changes were approved in issue #___
- [ ] No

### Checklist
- [ ] When resolving issues, they are referenced in the PR's title (e.g `fix: remove a typo, closes #___, #___`)
- [x] A change file is added if any packages will require a version bump due to this PR per [the instructions in the readme](https://github.com/tauri-apps/tauri/blob/dev/.changes/readme.md).
- [x] I have added a convincing reason for adding this feature, if necessary

### Other information

Currently the Tauri's `Window` type (both the Rust struct and JS class) are actually a mix of the `tao` Window and the `wry` Webview. In order to support multiple webviews on a given window, we actually have to split those types. This PR makes that change so now `WindowBuilder` and `Window` only exposes the window APIs, and we added `WebviewBuilder` and `Webview`.

To create a window with a webview attached to it, I've added [WindowBuilder::with_webview](https://github.com/tauri-apps/tauri/blob/41c2246d81b85f625aa3e8c3fb25b3bd95b0ed84/core/tauri/src/window/mod.rs#L265) on the Rust side and [WebviewWindow class](https://github.com/tauri-apps/tauri/blob/41c2246d81b85f625aa3e8c3fb25b3bd95b0ed84/tooling/api/src/webview.ts#L520) on the JS side. This reduces the breaking changes on the JS side (which I think it's the most common way of using multiwindow on Tauri).

To add webviews to a window, [Window::add_child](https://github.com/tauri-apps/tauri/blob/41c2246d81b85f625aa3e8c3fb25b3bd95b0ed84/core/tauri/src/window/mod.rs#L852) was added. The API isn't finished yet, so I propose we either hide it for v2 or keep it behind an `unstable` feature flag. Specially the auto resize thing needs more work and configuration.

Run `cargo run --example multiwebview` to play with it:
Screenshot 2023-11-21 at 16 57 36

Ref #2972 #2709 #2076

### CODE REVIEW

**Comment by lucasfernog** [comment]
We could also add a WebviewWindow struct on the Rust side if we want to minimize breaking changes, but that comes it a lot of duplication between the builder and window/webview types.

**Comment by amrbashir** [comment]
I think `WebviewWindow` is the correct way and I'd rather add it now than later because we already removed `WebviewWindow` in v2-alpha and would be annoying to bring it back in `v2.1` or `v3`

**Comment by ruoxiaomo** [comment]
I'm excited about this! I hope this feature can be released soon, as there is a high demand for it, including myself. My application urgently needs this functionality to meet the needs of my users.

**Comment by MessageDream** [comment]
截屏2023-11-29 15 17 59
I noticed in `core/tauri-runtime-wry/src/lib.rs` from line 2820 to 2831, there is code for adjusting the webview size along with the window size. I suggest considering the addition of a configuration option to control the execution of this code. If the configuration option is enabled, then execute this proportionally scaling code; if the configuration option is not enabled, then do not execute this code. The option could be enabled or disabled by default. This design can provide more flexibility to users, as sometimes we might not want to scale proportionally, such as the distance between two webviews, or the distance from the webview to the top or bottom.

**Comment by deminearchiver** [comment]
This feature is so underrated. It has a bunch of uses. I plan to use it to separate a custom titlebar from the content.

**Comment by Stuntlover-TM** [comment]
would love this to be in tauri

File: core/tauri-runtime/src/lib.rs [hunk 1]
```diff
@@ -194,13 +195,20 @@ pub trait RuntimeHandle<T: UserEvent>: Debug + Clone + Send + Sync + Sized + 'st
   /// Creates an `EventLoopProxy` that can be used to dispatch user events to the main event loop.
   fn create_proxy(&self) -> <Self::Runtime as Runtime<T>>::EventLoopProxy;
 
-  /// Create a new webview window.
+  /// Create a new window.
   fn create_window<F: Fn(RawWindow) + Send + 'static>(
     &self,
     pending: PendingWindow<T, Self::Runtime>,
-    before_webview_creation: Option<F>,
+    before_window_creation: Option<F>,
```
Comment (amrbashir) [inline core/tauri-runtime/src/lib.rs:217]: this new name is incorrect and should now be renamed `after_window_creation` or removed.

**Comment by boot2linux** [comment]
does this refactor support mobile platform ?

**Comment by lucasfernog** [comment]
mobile isn't supported on wry

File: core/tauri-runtime/src/lib.rs [hunk 2]
```diff
@@ -194,13 +195,20 @@ pub trait RuntimeHandle<T: UserEvent>: Debug + Clone + Send + Sync + Sized + 'st
...
   ) -> Result<DetachedWindow<T, Self::Runtime>>;
 
+  /// Create a new window.
```
Comment (amrbashir) [inline core/tauri-runtime/src/lib.rs]:
```suggestion
  /// Create a new webview.
```

File: core/tauri/src/menu/plugin.rs [hunk 3]
```diff
@@ -80,7 +80,7 @@ enum Predefined {
   Hide,
   HideOthers,
   ShowAll,
-  CloseWindow,
+  CloseWebview,
```
Comment (amrbashir) [inline core/tauri/src/menu/plugin.rs]: Is this intentional? it should stay `CloseWindow`

File: core/tauri/src/menu/plugin.rs [hunk 4]
```diff
@@ -769,7 +769,7 @@ fn set_accelerator<R: Runtime>(
 }
 
 #[command(root = "crate")]
-fn set_as_windows_menu_for_nsapp<R: Runtime>(
+fn set_as_webviews_menu_for_nsapp<R: Runtime>(
```
Comment (amrbashir) [inline core/tauri/src/menu/plugin.rs]: Another one

**Review by amr-crabnebula — APPROVED** [review]

**Review by amrbashir — APPROVED** [review]

**Comment by thewh1teagle** [comment]
Opened related issue
https://github.com/tauri-apps/tauri/issues/8676

**Comment by jkinggg** [comment]
Hi @lucasfernog, this is an awesome feature you've been working on. I've been playing around with it, and the only issue I've had is that it completely overlays the main tauri webview, so things can get hidden behind it. It makes things like managing drop downs and headers harder. Is it possible to position this webview so it sits under the main webview, enabling layering things on top?

**Comment by nileshtrivedi** [comment]
Is there an example of adding a new child webview to the main window when processing a command?

**Pull request merged**